server.go

```diff
@@ -83,6 +83,19 @@ type ServeConfig struct {
 	// Logger is used to pass a logger into the server. If none is provided the
 	// server will create a default logger.
 	Logger hclog.Logger
+
+	// Listener is the listener that the plugin server will listen for
+	// plugin connections. THIS DOES NOT NORMALLY NEED TO BE SET. If this
+	// isn't set, the plugin chooses a listener. This is exposed in case you
+	// want to carefully control how a plugin is served.
+	//
+	// If TLSProvider is set, this listener will be wrapped with a TLS
+	// listener. If you want to manually control TLS you should set
+	// TLSProvider to nil but be aware that the client side will need to be
+	// manually made aware of the certificate used.
+	//
+	// It is the callers responsibility to close this listener.
```

Comment:
This responsibility seemed fine on the surface, but I was trying to figure out what calling code might look like to deal with this and I feel a little uncertain... isn't the usual pattern that `Serve` just never returns and calls `os.Exit` if it wants to halt the process early? In which case, it seems like the caller wouldn't be able to know when it's time to close the listener.

Might it be clearer to document that the `Serve` function takes full ownership of this listener -- that the caller shouldn't do anything further with it, including closing it -- and then have `Serve` close it at the same time it would normally close a listener it created itself?

Reply:
Agreed done.